crates/miden-lib/src/note/well_known_note.rs

```diff
@@ -226,4 +232,234 @@ impl WellKnownNote {
+
+/// Returns the receiver account ID parsed from the provided P2ID note inputs.
+///
+/// # Errors
+///
+/// Returns an error if:
+/// - the length of the provided note inputs array is not equal to the expected inputs number of the
+///   P2ID note.
+/// - first two elements of the note inputs array does not form the valid account ID.
+fn parse_p2id_inputs(note_inputs: &[Felt]) -> Result<AccountId, StaticAnalysisError> {
+    if note_inputs.len() != WellKnownNote::P2ID.num_expected_inputs() {
+        return Err(StaticAnalysisError::new(format!(
+            "P2ID note should have {} inputs, but {} was provided",
+            WellKnownNote::P2ID.num_expected_inputs(),
+            note_inputs.len()
+        )));
+    }
+
+    try_read_account_id_from_inputs(note_inputs)
+}
+
+/// Returns the receiver account ID, reclaim height and timelock height parsed from the provided
+/// P2IDE note inputs.
+///
+/// # Errors
+///
+/// Returns an error if:
+/// - the length of the provided note inputs array is not equal to the expected inputs number of the
+///   P2IDE note.
+/// - first two elements of the note inputs array does not form the valid account ID.
+/// - third note inputs array element (reclaim height) is not a valid u32 value.
+/// - fourth note inputs array element (timelock height) is not a valid u32 value.
+fn parse_p2ide_inputs(note_inputs: &[Felt]) -> Result<(AccountId, u32, u32), StaticAnalysisError> {
+    if note_inputs.len() != WellKnownNote::P2IDE.num_expected_inputs() {
+        return Err(StaticAnalysisError::new(format!(
+            "P2IDE note should have {} inputs, but {} was provided",
+            WellKnownNote::P2IDE.num_expected_inputs(),
+            note_inputs.len()
+        )));
+    }
+
+    let receiver_account_id = try_read_account_id_from_inputs(note_inputs)?;
+
+    let reclaim_height = u32::try_from(note_inputs[2])
+        .map_err(|_err| StaticAnalysisError::new("reclaim block height should be a u32"))?;
+
+    let timelock_height = u32::try_from(note_inputs[3])
+        .map_err(|_err| StaticAnalysisError::new("timelock block height should be a u32"))?;
+
+    Ok((receiver_account_id, reclaim_height, timelock_height))
+}
+
+/// Reads the account ID from the first two note input values.
+///
+/// Returns None if the note input values used to construct the account ID are invalid.
+fn try_read_account_id_from_inputs(note_inputs: &[Felt]) -> Result<AccountId, StaticAnalysisError> {
+    if note_inputs.len() < 2 {
+        return Err(StaticAnalysisError::new(format!(
+            "P2ID and P2IDE notes should have at least 2 note inputs, but {} was provided",
+            note_inputs.len()
+        )));
+    }
+
+    let account_id_felts: [Felt; 2] = note_inputs[0..2].try_into().map_err(|source| {
+        StaticAnalysisError::with_source(
+            "should be able to convert the first two note inputs to an array of two Felt elements",
+            source,
+        )
+    })?;
```

Comment:
Why do we need this conversion? We should be able to just do `[note_inputs[1], note_inputs[0]]` in the code below.

Reply:
Ah, right, my bad. I'll remove it right away.